sound/soc/sof/intel/hda-dsp.c

```diff
@@ -477,26 +477,35 @@ static int hda_dsp_set_D0_state(struct snd_sof_dev *sdev,
 
 		if (hda_dsp_d0i3_streaming_applicable(sdev))
 			flags |= HDA_PM_PG_STREAMING;
+
+		ret = hda_dsp_send_pm_gate_ipc(sdev, flags);
+		if (ret < 0) {
+			dev_err(sdev->dev, "PM_GATE ipc error %d\n", ret);
+			return ret;
+		}
+
+		/* update D0I3C register */
+		ret = hda_dsp_update_d0i3c_register(sdev, value);
+		if (ret < 0)
+			return ret;
 	} else {
 		/* prevent power gating in D0I0 */
 		flags = HDA_PM_PPG;
-	}
 
-	/* update D0I3C register */
-	ret = hda_dsp_update_d0i3c_register(sdev, value);
-	if (ret < 0)
-		return ret;
+		ret = hda_dsp_update_d0i3c_register(sdev, value);
+		if (ret < 0)
+			return ret;
 
-	/*
-	 * Notify the DSP of the state change.
-	 * If this IPC fails, revert the D0I3C register update in order
-	 * to prevent partial state change.
-	 */
-	ret = hda_dsp_send_pm_gate_ipc(sdev, flags);
-	if (ret < 0) {
-		dev_err(sdev->dev,
-			"error: PM_GATE ipc error %d\n", ret);
-		goto revert;
+		/*
+		 * Notify the DSP of the state change.
+		 * If this IPC fails, revert the D0I3C register update in order
+		 * to prevent partial state change.
+		 */
+		ret = hda_dsp_send_pm_gate_ipc(sdev, flags);
+		if (ret < 0) {
+			dev_err(sdev->dev, "PM_GATE ipc error %d\n", ret);
+			goto revert;
+		}
```

Comment:
ok, so I am going to put the emergency brake on this one.

a) we've enable D0i3-WoV for TGL and the code on the left worked fine.
b) the code on the right uses a different sequence for D0i3 entry and exit. That would mean D0i3 as we know it never worked?

One of these two propositions is problematic. I don't have a clue on what the sequence should be, just that I can only deal with very binary logic.

@ranj063 Does this look right to you?

Reply:
@plbossart I think this change is logical as it makes entry/exit symmetrical. But I have a feeling that it worked for us before it doesnt really matter.

Reply:
that's what I am trying to get to @ranj063, does the order matter at all?

Reply:
I dont think it matters up until ADL as we have tested it to work. But IM not sure for MTL. @RanderWang can you confirm?

Reply:
@ranj063  In my test the different order get the same result on MTL. Since pmc driver on mtl is not ready so we can't make any conclusion.  Two month ago, I discussed with keyon about d0i3 register setting in host and have a same idea that:  d0i3 register is only a notification for pmc, not a switch to enable or disable.  According to spec, the power saving is mainly done in audio hw component, not by pmc. It is possible that the order doesn't matter even pmc driver is ready.  But the order in this PR is comfortable to me.